apps/src/dance/SongSelector.tsx

```diff
@@ -0,0 +1,110 @@
+import React, {useState, useCallback, ChangeEvent} from 'react';
+import classnames from 'classnames';
+
+import {commands as audioCommands} from '@cdo/apps/lib/util/audioApi';
+import {getFilteredSongKeys} from '@cdo/apps/dance/songs';
+import Button from '@cdo/apps/templates/Button';
+import {SongData} from '@cdo/apps/dance/types';
+import moduleStyles from '@cdo/apps/dance/song-selector.module.scss';
+
+const commonI18n = require('@cdo/locale');
+
+const currentTimeoutsMap: {[key: string]: ReturnType<typeof setTimeout>} = {};
+
+interface SongSelectorProps {
+  selectedSong: string;
+  levelIsRunning?: boolean;
+  setSong: (songId: string) => void;
+  songData: SongData;
+  enableSongSelection?: boolean;
+  filterOn?: boolean;
+}
+
+const SongSelector: React.FC<SongSelectorProps> = ({
+  selectedSong,
+  levelIsRunning,
+  setSong,
+  songData,
+  enableSongSelection,
+  filterOn,
+}) => {
+  const [songInPreview, setSongInPreview] = useState(false);
+
+  const onPreviewBtnClick = useCallback(() => {
+    if (songInPreview) {
+      audioCommands.stopSound({url: songData[selectedSong].url});
+      setSongInPreview(false);
+    } else {
+      audioCommands.playSound({
+        url: `${songData[selectedSong].url}`,
+        callback: () => {
+          setSongInPreview(true);
+          const timeoutID = setTimeout(() => {
+            if (!levelIsRunning) {
+              audioCommands.stopSound({url: songData[selectedSong].url});
+              setSongInPreview(false);
+            }
+          }, 10000);
+
+          currentTimeoutsMap[selectedSong] = timeoutID;
+        },
+        onEnded: () => {
+          currentTimeoutsMap[selectedSong] &&
+            clearTimeout(currentTimeoutsMap[selectedSong]);
+
+          delete currentTimeoutsMap[selectedSong];
+
+          setSongInPreview(false);
+        },
+      });
+    }
+  }, [levelIsRunning, selectedSong, songData, songInPreview]);
+
+  const songKeys = getFilteredSongKeys(songData, filterOn);
+  const changeSong = useCallback(
```

Comment:
For my understanding - in your video, it looks like when you change songs mid-preview, playback stops (which makes sense). Where does that happen here? I wasn't sure since I didn't see a call to `setSongInPreview(false)` on song change, but maybe it's handled another way?

Reply:
Got it, thanks for the explanation!